link/src/main/java/com/stripe/android/link/ui/signup/SignUpViewModel.kt

```diff
@@ -0,0 +1,225 @@
+        value = SignUpScreenState.Loading
+    )
+
+    val state: StateFlow<SignUpScreenState> = _state
+
+    val requiresNameCollection: Boolean
+        get() {
+            val countryCode = when (val stripeIntent = args.configuration.stripeIntent) {
+                is PaymentIntent -> stripeIntent.countryCode
+                is SetupIntent -> stripeIntent.countryCode
+            }
+            return countryCode != CountryCode.US.value
+        }
+
+    init {
+        viewModelScope.launch {
+            loadScreen()
+        }
+        viewModelScope.launch {
+            signUpStateListener()
+        }
+        viewModelScope.launch {
+            emailListener()
+        }
+        linkEventsReporter.onSignupFlowPresented()
+    }
+
+    private suspend fun loadScreen() {
+        val isLoggedOut = linkAccountManager.hasUserLoggedOut(args.configuration.customerInfo.email)
+        val newState = SignUpScreenState.Content(
+            emailController = EmailConfig.createController(
+                initialValue = args.configuration.customerInfo.email.takeUnless { isLoggedOut }
+            ),
+            phoneNumberController = PhoneNumberController.createPhoneNumberController(
+                initialValue = args.configuration.customerInfo.phone.takeUnless { isLoggedOut }.orEmpty(),
+                initiallySelectedCountryCode = args.configuration.customerInfo.billingCountryCode,
+            ),
+            nameController = NameConfig.createController(
+                initialValue = args.configuration.customerInfo.name.takeUnless { isLoggedOut }
+            ),
+            signUpEnabled = false
+        )
+        _state.emit(newState)
+    }
+
+    private suspend fun signUpStateListener() {
+        _state.flatMapLatest { state ->
+            if (state !is SignUpScreenState.Content) return@flatMapLatest flowOf()
+            combine(
+                flow = state.nameController.fieldState.map {
+                    if (requiresNameCollection) {
+                        it.isValid()
+                    } else {
+                        true
+                    }
+                },
+                flow2 = state.emailController.fieldState.map { it.isValid() },
+                flow3 = state.phoneNumberController.isComplete
+            ) { nameComplete, emailComplete, phoneComplete ->
+                nameComplete && emailComplete && phoneComplete
+            }
+        }.collectLatest { formValid ->
+            updateSignUpEnabled(formValid)
+        }
+    }
+
+    private suspend fun emailListener() {
+        _state.flatMapLatest { state ->
+            if (state !is SignUpScreenState.Content) return@flatMapLatest flowOf()
+            state.emailController.formFieldValue.mapLatest { entry ->
+                entry.takeIf { it.isComplete }?.value
+            }
+        }.collectLatest { email ->
+            delay(LOOKUP_DEBOUNCE)
+            if (email != null) {
```

Comment:
Collection block will be cancelled and replaced when new email is emitted. Debounce is meant to delay lookup in case new email gets emitted quickly. [Old version](https://github.com/stripe/stripe-android/blob/1a7c04ae93d78d8c5c18e6eba13e7fd8a149a55d/link/src/main/java/com/stripe/android/link/ui/signup/SignUpViewModel.kt#L68) used a debouncer as well.